### PR TITLE
feat: View model processors now work on POJOs and documentation improvements

### LINKED ISSUE #229: Document ViewModelProcessor
Its mentioned in relation to security, but no specifics docs about how to register your own

### LINKED ISSUE #247: POJOs passed to views are not being processed by ViewModelProcessors
@rferreira 

> ModelViewProcessors are not applied to non-map like models due to its strict Map<String, Object> interface. We should consider revisiting this in favor of a <T> interface for v3

Context for this issue:

https://github.com/micronaut-projects/micronaut-views/pull/219  and: https://github.com/micronaut-projects/micronaut-views/pull/246

Maybe we can use [BeanIntrospection](https://docs.micronaut.io/latest/guide/#introspection) to support processing Models which are POJOs annotated with `@Introspected`

### PR DESCRIPTION
**Pull request by rferreira**
* This includes a breaking change to ViewModelProcessor.java
* Grouped all template engines related documentation into a templates subsection
* Grouped all security related pages into a security subsection
* Added documentation page on enhancing models using a custom processors

Fixes #229
Fixes #247

### CODE REVIEW

**Comment by rferreira** [comment]
Sorry for the large PR here folks but most of it is shuffling the template docs into a templates subsection.

File: views-core/src/main/java/io/micronaut/views/ViewsFilter.java [hunk 1]
```diff
@@ -75,107 +71,63 @@ public int getOrder() {
                                                             ServerFilterChain chain) {
 
         return Flux.from(chain.proceed(request))
-            .switchMap(response -> {
-                Optional<AnnotationMetadata> routeMatch = response.getAttribute(HttpAttributes.ROUTE_MATCH,
-                        AnnotationMetadata.class);
-                if (routeMatch.isPresent()) {
-                    AnnotationMetadata route = routeMatch.get();
+                .switchMap(response -> {
```
Comment (rferreira) [inline views-core/src/main/java/io/micronaut/views/ViewsFilter.java:74]: I did a medium size refactor of the ViewFilter, putting most exits at the top of the method and simplifying the logic (by removing as much of it as possible)

File: views-core/src/main/java/io/micronaut/views/ViewsRenderer.java [hunk 2]
```diff
@@ -37,39 +38,44 @@
     String EXTENSION_SEPARATOR = ".";
 
     /**
-     * @param viewName view name to be render
+     * @param viewName view name to be rendered
```
Comment (rferreira) [inline views-core/src/main/java/io/micronaut/views/ViewsRenderer.java:41]: edits here are minor, just some language cleanup

File: views-core/src/main/java/io/micronaut/views/ViewsRenderer.java [hunk 3]
```diff
@@ -37,39 +38,44 @@
...
     /**
-     * Creates a view model for the given data.
+     * Converts the object being passed in into a {@code Map<String, Object>} if not already one.
+     *
      * @param data The data
      * @return The model
      */
-    default @NonNull Map<String, Object> modelOf(@Nullable Object data) {
+    default @NonNull
+    Map<String, Object> modelOf(@Nullable Object data) {
```
Comment (rferreira) [inline views-core/src/main/java/io/micronaut/views/ViewsRenderer.java:73]: I contemplated removing this method but I decided for keeping it to minimize the backwards breaking changes

File: views-core/src/main/java/io/micronaut/views/model/ViewModelProcessor.java [hunk 4]
```diff
@@ -15,27 +15,26 @@
...
- * Implementers of {@link ViewModelProcessor} process the {@link ModelAndView} and modify it prior to rendering by
- * either adding or removing entries.
+ * Implementers of {@link ViewModelProcessor} process the {@link ModelAndView} and modify it prior to rendering.
  *
+ * @param <T> the type this model processor should apply to
  * @author Sergio del Amo
  * @since 1.0
  */
-public interface ViewModelProcessor {
+public interface ViewModelProcessor<T> {
```
Comment (rferreira) [inline views-core/src/main/java/io/micronaut/views/model/ViewModelProcessor.java:29]: Here lies the biggest change, ViewModelProcessors now are typed

File: views-core/src/main/java/io/micronaut/views/model/security/SecurityViewModelProcessor.java [hunk 5]
```diff
@@ -30,15 +30,15 @@
...
+ * Returns information about the current user so that it can be appended it to the model being rendered.
  *
  * @author Sergio del Amo
  * @since 1.1.0
  */
 @Requires(property = SecurityViewModelProcessorConfigurationProperties.PREFIX + ".enabled", notEquals = StringUtils.FALSE)
 @Requires(beans = {SecurityFilter.class, SecurityService.class, SecurityViewModelProcessorConfiguration.class})
 @Singleton
-public class SecurityViewModelProcessor implements ViewModelProcessor {
+public class SecurityViewModelProcessor implements ViewModelProcessor<Map<String, Object>> {
```
Comment (rferreira) [inline views-core/src/main/java/io/micronaut/views/model/security/SecurityViewModelProcessor.java:41]: Updating existing processors to the new type is as simple as setting the type of Map<String, Object>

**Review by sdelamo — CHANGES_REQUESTED** [review]

**Comment by sdelamo** [comment]
Thanks for the contribution @rferreira I have moved the contents of this PR to several PRs:

-  https://github.com/micronaut-projects/micronaut-views/pull/263
- https://github.com/micronaut-projects/micronaut-views/pull/262

**Comment by sdelamo** [comment]
@rferreira With the above linked PRs and #265 I think it covers everything suggested by this PR. I am going to close this PR.

Please, leave any comments at #265 at let me know if you have any concerns or you think I did not move any necessary changes.

**Pull request closed without merging**